platform/qt/src/qquickmapboxgl.cpp

```diff
@@ -182,132 +203,254 @@ void QQuickMapboxGL::pan(int dx, int dy)
-    m_style = style;
-    if (style) {
-        style->setParentItem(this);
-        connect(style, SIGNAL(urlChanged(QString)), this, SLOT(onStyleChanged()));
-        connect(style, SIGNAL(propertyUpdated(QVariantMap)), this, SLOT(onStylePropertyUpdated(QVariantMap)));
+    // Reload parameters if switching style URLs.
+    if (!m_styleUrl.isEmpty()) {
+        for (const auto& param_ : m_parameters) {
+            if (param_->property("type").toString() == "style") continue;
+            processMapParameter(param_);
+        }
     }
 
+    m_styleUrl = url;
+    m_styleLoaded = false;
     m_syncState |= StyleNeedsSync;
-    update();
 
-    emit styleChanged();
+    return true;
 }
 
-QQuickMapboxGLStyle *QQuickMapboxGL::style() const
+bool QQuickMapboxGL::parseStyleProperty(QQuickMapboxGLMapParameter *param, const QString &name)
 {
-    return m_style;
-}
+    // Ignore meta-properties "type", "layer" and "class".
+    if (name == "type" || name == "layer" || name == "class") {
+        return false;
+    }
 
-void QQuickMapboxGL::setBearing(qreal angle)
-{
-    angle = std::fmod(angle, 360.);
+    QString formattedName(name);
+    formattedName = formattedName.replace(s_camelCase, "\\1-\\2").toLower();
 
-    if (m_bearing == angle) {
-        return;
+    QVariant value = param->property(name.toLatin1());
+    if (value.canConvert<QJSValue>()) {
+        value = value.value<QJSValue>().toVariant();
     }
 
-    m_bearing = angle;
+    m_stylePropertyChanges << QQuickMapboxGL::StyleProperty {
+        param->property("type").toString().at(0) == 'p' ? StyleProperty::Paint : StyleProperty::Layout,
+        param->property("layer").toString(),
+        formattedName,
+        value,
+        param->property("class").toString()
+    };
+    return true;
+}
 
-    m_syncState |= BearingNeedsSync;
-    update();
+bool QQuickMapboxGL::parseStyleLayer(QQuickMapboxGLMapParameter *param)
+{
+    QVariantMap layer;
+    layer["id"] = param->property("name");
+    layer["source"] = param->property("source");
+    layer["type"] = param->property("layerType");
+    if (param->property("sourceLayer").isValid()) {
+        layer["source-layer"] = param->property("sourceLayer");
+    }
+    m_layerChanges << layer;
+    return true;
+}
+
+bool QQuickMapboxGL::parseStyleSource(QQuickMapboxGLMapParameter *param)
+{
+    QString sourceType = param->property("sourceType").toString();
+
+    QVariantMap source;
+    source["id"] = param->property("name");
+    source["type"] = sourceType;
+    if (sourceType == "vector" || sourceType == "raster") {
+        source["url"] = param->property("url");
+        m_sourceChanges << source;
+    } else if (sourceType == "geojson") {
+        QFile geojson(param->property("data").toString());
+        geojson.open(QIODevice::ReadOnly);
+        source["data"] = geojson.readAll();
+        m_sourceChanges << source;
+    } else {
+        qWarning() << "Error: Unsupported source type '" << sourceType << "'.";
```

Comment:
I see this a nice fit to the error API we inherited from the QML Map. Also I find this message a bit confusing for the end user.

```
Error { NoError, NotSupportedError, UnknownParameterError, MissingRequiredParameterError, ConnectionError }
```